wit/wasi-nn.wit

```diff
@@ -0,0 +1,100 @@
+package wasi:nn
+
+world inference {
+    import tensor
+    import graph
+    import execution
+    import errors
+
+    export run: func()-> result<_,string>
+}
+
+interface tensor {
+    use errors.{error}
+
+    type tensor-dimensions = list<u32>
+    type tensor-data = list<u8>
+
+    enum tensor-type {
+        fp16,
+        fp32,
+        bf16,
+        up8,
+        ip32
+    }
+
+    record tensor {
+        // Describe the size of the tensor (e.g., 2x2x2x2 -> [2, 2, 2, 2]). To represent a tensor
+        // containing a single value, use `[1]` for the tensor dimensions.
+        dimensions: tensor-dimensions,
+
+        // Describe the type of element in the tensor (e.g., f32).
+        tensor-type: tensor-type,
+
+        // Contains the tensor data.
+        data: tensor-data,
+    }
+}
+
+interface graph {
+    use errors.{error}
+    type graph-builder = list<u8>
+    type graph-builder-array = list<graph-builder>
+    use tensor.{tensor}
+
+    type graph = u32
+
+    enum graph-encoding {
+        openvino,
+        onnx,
+        tensorflow,
+        pytorch,
+        tensorflowlite
+    }
+
+    enum execution-target {
+        cpu,
+        gpu,
+        tpu
+    }
+
+    load: func(builder: graph-builder-array, encoding: graph-encoding, target: execution-target) -> result<graph, error>
+    load-named-model: func(name: string) -> result<graph, error>
+}
+
+interface execution {
+    use errors.{error}
+    use tensor.{tensor, tensor-data}
+    use graph.{graph}
+
+    type graph-execution-context = u32
+    init-execution-context: func(graph: graph) -> result<graph-execution-context, error>
+    set-input: func(ctx: graph-execution-context, index: u32, tensor: tensor) -> result<_, error>
+    compute: func(ctx: graph-execution-context) -> result<_, error>
+    get-output: func(ctx: graph-execution-context, index: u32) -> result<list<tensor-data>, error>
+    eval: func(tensors: list<tensor>) -> result<list<tensor>, error>
+    
+}
+
+interface errors {
+    enum error {
+        // No error occurred.
+        success,
+        // Caller module passed an invalid argument.
+        invalid-argument,
+        // Invalid encoding.
+        invalid-encoding,
+        // Caller module is missing a memory export.
+        missing-memory,
+        // Device or resource busy.
+        busy,
+        // Runtime Error.
+        runtime-error,
+        // Unsupported operation
+        unsupported-operation,
+        // Model too large
+        model-too-large,
+        // Model not found
+        model-not-found
```

Comment:
these last two variants, `model-too-large` and `model-not-found`, i expect are particular to the `load` and `load-named-model` methods. can you instead define an enum of errors particular to those two methods in the `graph` interface and move this definition to the `execution` interface?

Reply:
So unsupported operation and invalid argument are currently shared between execution and graph interfaces. I'm not sure about busy and runtime error (@abrown might know), but invalid-encoding, model-too-large, model-not-found are all graph specific errors. The only errors that can happen during execution are 
- tensor data type or dimension mismatch with graph
- framework does not support setting input by name 
- framework does not support setting input by index
- index out of range when setting input or getting output
- Getting NaN or +/- inf in the computation graph (very unlikely, but tends to crash execution, guessing this corresponds to runtime error)

We can introduce additional execution error and graph error enums here, but @abrown may have some thoughts about the size of that change.